In [1]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
# Import data by extracting 600  primes in dataset primes.txt
# Initialize an empty list to hold the elements
elements_list = []

# Open the file in read mode
with open('/content/drive/MyDrive/primes.txt', 'r') as file:
    # Use enumerate to get both line numbers and lines
    for i, line in enumerate(file):

        elements_list.append(line.strip())  # strip() removes any leading/trailing whitespace, including newlines
        # Break the loop
        if i == 600:
            break

my_list = [int(item) for item in elements_list]


In [3]:
# Generate the product of primes with themselves and then sort the products in descending order
products = [a * b for a in my_list for b in my_list]
results = list(dict.fromkeys(products))
# sort the result
sorted_results = sorted(results)


#len(sorted_results)

In [4]:
# The length of our data after multiplying prime numbers with themselves
len(sorted_results)

180901

In [5]:
# Check for what is the maxium number of the products
max(sorted_results)

19545241

In [6]:
# This function is to divide the data of product of primes into equivalent chunks. Dividing dataset into smaller chunks helps our algorithm run faster

def divide_input(input_list):
    # The total length of the input list
    total_length = len(input_list)

    # The size of each chunk
    chunk_size = total_length // 10

    # List to hold all the chunks
    chunks = []

    for i in range(0, total_length, chunk_size):
        # Append chunks to the list
        chunks.append(input_list[i:i + chunk_size])

    return chunks

In [7]:
chunks = divide_input(sorted_results)

 Each json file stores the tuple whose difference between 2 numbers is 1, i.e tuple(n, n+1) . Each file has the length is 5001500

In [8]:
import json
with open('/content/drive/MyDrive/data_v2/tuple_diff1_chunk_1.json', 'r') as file:
    loaded_data = json.load(file)
restored_tuples_diff1 = [tuple(item) for item in loaded_data]


In [9]:
len(restored_tuples_diff1)

5001500

In [10]:
# First 10 rows of data
restored_tuples_diff1[:10]

[(1, 2),
 (2, 3),
 (3, 4),
 (4, 5),
 (5, 6),
 (6, 7),
 (7, 8),
 (8, 9),
 (9, 10),
 (10, 11)]

In [11]:
# Last 10 rows of data
restored_tuples_diff1[-10:]

[(5001491, 5001492),
 (5001492, 5001493),
 (5001493, 5001494),
 (5001494, 5001495),
 (5001495, 5001496),
 (5001496, 5001497),
 (5001497, 5001498),
 (5001498, 5001499),
 (5001499, 5001500),
 (5001500, 5001501)]

In [12]:
with open('/content/drive/MyDrive/data_v2/tuple_diff1_chunk_2.json', 'r') as file:
    loaded_data = json.load(file)
restored_tuples_diff1_v1 = [tuple(item) for item in loaded_data]

In [13]:
restored_tuples_diff1_v1[:10]

[(5001501, 5001502),
 (5001502, 5001503),
 (5001503, 5001504),
 (5001504, 5001505),
 (5001505, 5001506),
 (5001506, 5001507),
 (5001507, 5001508),
 (5001508, 5001509),
 (5001509, 5001510),
 (5001510, 5001511)]

In [14]:
restored_tuples_diff1_v1[-10:]

[(10002991, 10002992),
 (10002992, 10002993),
 (10002993, 10002994),
 (10002994, 10002995),
 (10002995, 10002996),
 (10002996, 10002997),
 (10002997, 10002998),
 (10002998, 10002999),
 (10002999, 10003000),
 (10003000, 10003001)]

In [15]:
with open('/content/drive/MyDrive/data_v2/tuple_diff1_chunk_3.json', 'r') as file:
    loaded_data = json.load(file)
restored_tuples_diff1_v2 = [tuple(item) for item in loaded_data]

In [16]:
restored_tuples_diff1_v2[:10]

[(10003001, 10003002),
 (10003002, 10003003),
 (10003003, 10003004),
 (10003004, 10003005),
 (10003005, 10003006),
 (10003006, 10003007),
 (10003007, 10003008),
 (10003008, 10003009),
 (10003009, 10003010),
 (10003010, 10003011)]

In [17]:
restored_tuples_diff1_v2[-10:]

[(15004491, 15004492),
 (15004492, 15004493),
 (15004493, 15004494),
 (15004494, 15004495),
 (15004495, 15004496),
 (15004496, 15004497),
 (15004497, 15004498),
 (15004498, 15004499),
 (15004499, 15004500),
 (15004500, 15004501)]

In [18]:
with open('/content/drive/MyDrive/data_v2/tuple_diff1_chunk_4.json', 'r') as file:
    loaded_data = json.load(file)
restored_tuples_diff1_v3 = [tuple(item) for item in loaded_data]

In [19]:
restored_tuples_diff1_v3[:10]

[(15004501, 15004502),
 (15004502, 15004503),
 (15004503, 15004504),
 (15004504, 15004505),
 (15004505, 15004506),
 (15004506, 15004507),
 (15004507, 15004508),
 (15004508, 15004509),
 (15004509, 15004510),
 (15004510, 15004511)]

In [20]:
restored_tuples_diff1_v3[-10:]

[(20005991, 20005992),
 (20005992, 20005993),
 (20005993, 20005994),
 (20005994, 20005995),
 (20005995, 20005996),
 (20005996, 20005997),
 (20005997, 20005998),
 (20005998, 20005999),
 (20005999, 20006000),
 (20006000, 20006001)]

In [21]:
# Combine 3 small dataset above and we obtain a new one named tuple_diff1 whose length is 2000600
tuple_diff1 = restored_tuples_diff1 + restored_tuples_diff1_v1 +restored_tuples_diff1_v2 + restored_tuples_diff1_v3

In [22]:
len(tuple_diff1)

20006000


 Each json file stores the tuple whose difference between 2 numbers is 2, i.e tuple(n, n+2) . Each file has the length is 5001499

In [23]:
with open('/content/drive/MyDrive/data_v2/tuple_diff2_chunk_1.json', 'r') as file:
    loaded_data = json.load(file)
    restored_tuples_diff2 = [tuple(item) for item in loaded_data]


In [24]:
len(restored_tuples_diff2)

5001499

In [25]:
restored_tuples_diff2[:10]

[(1, 3),
 (2, 4),
 (3, 5),
 (4, 6),
 (5, 7),
 (6, 8),
 (7, 9),
 (8, 10),
 (9, 11),
 (10, 12)]

In [26]:
restored_tuples_diff2[-10:]

[(5001490, 5001492),
 (5001491, 5001493),
 (5001492, 5001494),
 (5001493, 5001495),
 (5001494, 5001496),
 (5001495, 5001497),
 (5001496, 5001498),
 (5001497, 5001499),
 (5001498, 5001500),
 (5001499, 5001501)]

In [27]:
with open('/content/drive/MyDrive/data_v2/tuple_diff2_chunk_2.json', 'r') as file:
    loaded_data = json.load(file)
    restored_tuples_diff2_v1 = [tuple(item) for item in loaded_data]

In [28]:
restored_tuples_diff2_v1[:10]

[(5001501, 5001503),
 (5001502, 5001504),
 (5001503, 5001505),
 (5001504, 5001506),
 (5001505, 5001507),
 (5001506, 5001508),
 (5001507, 5001509),
 (5001508, 5001510),
 (5001509, 5001511),
 (5001510, 5001512)]

In [29]:
restored_tuples_diff2_v1[-10:]

[(10002990, 10002992),
 (10002991, 10002993),
 (10002992, 10002994),
 (10002993, 10002995),
 (10002994, 10002996),
 (10002995, 10002997),
 (10002996, 10002998),
 (10002997, 10002999),
 (10002998, 10003000),
 (10002999, 10003001)]

In [30]:
with open('/content/drive/MyDrive/data_v2/tuple_diff2_chunk_3.json', 'r') as file:
    loaded_data = json.load(file)
    restored_tuples_diff2_v2 = [tuple(item) for item in loaded_data]

In [31]:
restored_tuples_diff2_v2[:10]

[(10003001, 10003003),
 (10003002, 10003004),
 (10003003, 10003005),
 (10003004, 10003006),
 (10003005, 10003007),
 (10003006, 10003008),
 (10003007, 10003009),
 (10003008, 10003010),
 (10003009, 10003011),
 (10003010, 10003012)]

In [32]:
restored_tuples_diff2_v2[-10:]

[(15004490, 15004492),
 (15004491, 15004493),
 (15004492, 15004494),
 (15004493, 15004495),
 (15004494, 15004496),
 (15004495, 15004497),
 (15004496, 15004498),
 (15004497, 15004499),
 (15004498, 15004500),
 (15004499, 15004501)]

In [33]:
with open('/content/drive/MyDrive/data_v2/tuple_diff2_chunk_4.json', 'r') as file:
    loaded_data = json.load(file)
    restored_tuples_diff2_v3 = [tuple(item) for item in loaded_data]

In [34]:
restored_tuples_diff2_v3[:10]

[(15004501, 15004503),
 (15004502, 15004504),
 (15004503, 15004505),
 (15004504, 15004506),
 (15004505, 15004507),
 (15004506, 15004508),
 (15004507, 15004509),
 (15004508, 15004510),
 (15004509, 15004511),
 (15004510, 15004512)]

In [35]:
restored_tuples_diff2_v3[-10:]

[(20005990, 20005992),
 (20005991, 20005993),
 (20005992, 20005994),
 (20005993, 20005995),
 (20005994, 20005996),
 (20005995, 20005997),
 (20005996, 20005998),
 (20005997, 20005999),
 (20005998, 20006000),
 (20005999, 20006001)]

In [36]:
#Combine 3 small dataset above and we obtain a new one named tuple_diff1 whose length is 2000594
tuple_diff2 = restored_tuples_diff2 + restored_tuples_diff2_v1 + restored_tuples_diff2_v2 + restored_tuples_diff2_v3

In [37]:
len(tuple_diff2)

20005996

In [38]:
def process_chunk(chunk, chunk1_diff1, chunk1_diff2):
    # Sort the chunk to ensure the calculations are based on sorted order
    sorted_results = sorted(chunk)

    # Process for difference of 1
    tuple_diff1 = chunk1_diff1
    # Extract the numbers from our product of primes which have the difference is 1
    product_diff1 = tuple((sorted_results[i], sorted_results[i + 1]) for i in range(len(sorted_results) - 1) if sorted_results[i + 1] - sorted_results[i] == 1)
    set_tuple_diff1 = set(tuple_diff1)
    set_product_diff1 = set(product_diff1)
    # Find the common elements which are in both set_tuple_diff1 and set_product_diff1
    common_elements_diff1 = set_tuple_diff1.intersection(set_product_diff1)
    # The total number of common elements
    num_tuple_diff1 = len(common_elements_diff1)

    # Process for difference of 2
    tuple_diff2 = chunk1_diff2
     # Extract the numbers from our product of primes which have the difference is 1
    product_diff2 = tuple((sorted_results[i], sorted_results[i + 1]) for i in range(len(sorted_results) - 1) if sorted_results[i + 1] - sorted_results[i] == 2)
    set_tuple_diff2 = set(tuple_diff2)
    set_product_diff2 = set(product_diff2)
    # Find the common elements which are in both set_tuple_diff2 and set_product_diff2
    common_elements_diff2 = set_tuple_diff2.intersection(set_product_diff2)
    # The total number of common elements
    num_tuple_diff2 = len(common_elements_diff2)

    # Return the sums of common elements with the difference of 1 and 2
    return num_tuple_diff1, num_tuple_diff2, common_elements_diff1, common_elements_diff2

In [41]:
num_tuples_diff1, num_tuples_diff2, common_elements_diff1, common_elements_diff2 = process_chunk(sorted_results, tuple_diff1,  tuple_diff2)

In [42]:
# The total number of tuples whose difference is 1
num_tuples_diff1

423

In [43]:
# The total number of tuples whose difference is 1
num_tuples_diff2

6863

In [44]:
# The number of tuples whose difference is 1
#sum_tuple_diff1_v1

In [45]:
# The number of tuples whose difference is 2
#sum_tuple_diff2_v1

In [46]:
# Print out the tuples whose difference is 1 in a descending order
sorted(common_elements_diff1)

[(9, 10),
 (14, 15),
 (21, 22),
 (25, 26),
 (33, 34),
 (34, 35),
 (38, 39),
 (57, 58),
 (85, 86),
 (86, 87),
 (93, 94),
 (94, 95),
 (118, 119),
 (121, 122),
 (122, 123),
 (133, 134),
 (141, 142),
 (142, 143),
 (145, 146),
 (158, 159),
 (177, 178),
 (201, 202),
 (202, 203),
 (205, 206),
 (213, 214),
 (214, 215),
 (217, 218),
 (218, 219),
 (253, 254),
 (298, 299),
 (301, 302),
 (302, 303),
 (326, 327),
 (334, 335),
 (361, 362),
 (381, 382),
 (393, 394),
 (394, 395),
 (445, 446),
 (446, 447),
 (453, 454),
 (481, 482),
 (501, 502),
 (514, 515),
 (526, 527),
 (537, 538),
 (542, 543),
 (553, 554),
 (565, 566),
 (622, 623),
 (633, 634),
 (634, 635),
 (694, 695),
 (697, 698),
 (698, 699),
 (706, 707),
 (717, 718),
 (745, 746),
 (766, 767),
 (778, 779),
 (793, 794),
 (802, 803),
 (817, 818),
 (841, 842),
 (842, 843),
 (865, 866),
 (878, 879),
 (898, 899),
 (913, 914),
 (921, 922),
 (922, 923),
 (933, 934),
 (958, 959),
 (973, 974),
 (1006, 1007),
 (1041, 1042),
 (1042, 1043),
 (1046, 1047),
 (1

In [49]:
# Print out the tuples whose difference is 2 in a descending order
sorted(common_elements_diff2)

[(4, 6),
 (49, 51),
 (55, 57),
 (91, 93),
 (119, 121),
 (143, 145),
 (159, 161),
 (183, 185),
 (185, 187),
 (203, 205),
 (215, 217),
 (219, 221),
 (235, 237),
 (247, 249),
 (265, 267),
 (287, 289),
 (289, 291),
 (299, 301),
 (303, 305),
 (319, 321),
 (321, 323),
 (327, 329),
 (339, 341),
 (391, 393),
 (411, 413),
 (413, 415),
 (415, 417),
 (451, 453),
 (469, 471),
 (471, 473),
 (515, 517),
 (517, 519),
 (527, 529),
 (533, 535),
 (535, 537),
 (543, 545),
 (551, 553),
 (579, 581),
 (581, 583),
 (589, 591),
 (667, 669),
 (669, 671),
 (679, 681),
 (685, 687),
 (687, 689),
 (695, 697),
 (721, 723),
 (753, 755),
 (779, 781),
 (789, 791),
 (791, 793),
 (813, 815),
 (815, 817),
 (849, 851),
 (869, 871),
 (893, 895),
 (899, 901),
 (949, 951),
 (959, 961),
 (993, 995),
 (1055, 1057),
 (1057, 1059),
 (1077, 1079),
 (1079, 1081),
 (1099, 1101),
 (1119, 1121),
 (1133, 1135),
 (1135, 1137),
 (1139, 1141),
 (1145, 1147),
 (1147, 1149),
 (1157, 1159),
 (1165, 1167),
 (1167, 1169),
 (1189, 1191),
 (120

In [51]:
# The largest tuple
sorted(common_elements_diff2)[-1]

(17131319, 17131321)